Calculate the mean, median, mode and standard deviation for the problem
statements 1& 2.

Problem Statement 1:

The marks awarded for an assignment set for a Year 8 class of 20 students were as follows:

6 7 5 7 7 8 7 6 9 7 4 10 6 8 8 9 5 6 4 8

In [49]:
from scipy import stats
import numpy as np
import pandas as pd
import statistics as st
import math

 Formulas for calculating these are as follows:
 
 N : sample size
 
 mean = (1/N) * $\sum_{i=0}^N x_i$
 
 median:
 
 Case I) When n is odd:
 
 median =  $x_i$
 
 Case II) When n is even:
 
 median = ($x_i$ + $x_{i+1}$)/2
 
 mean: value that occurs the most frequently
 
 variance ($\sigma^{2}$) = (1/N) * ${\sum_{i=0}^N ({x_i - \bar x}})^2$ 
 
 standard deviation ($\sigma$) = $\sqrt{variance}$


In [32]:
data = [6, 7, 5, 7, 7, 8, 7, 6, 9, 7, 4, 10, 6, 8, 8, 9, 5, 6, 4, 8]
# sum = summation of data = 137
n = 20
sum = 137
mean = sum /n
mean

6.85

In [33]:
# mode : arrange the data in ascending order
data.sort()
data

[4, 4, 5, 5, 6, 6, 6, 6, 7, 7, 7, 7, 7, 8, 8, 8, 8, 9, 9, 10]

### mode = 7, as frequency of occrrance is 5

In [34]:
# using stats formula
stats.mode(data)

ModeResult(mode=array([7]), count=array([5]))

### Median 

In [35]:
data.sort()
data

[4, 4, 5, 5, 6, 6, 6, 6, 7, 7, 7, 7, 7, 8, 8, 8, 8, 9, 9, 10]

In [45]:
mid = (int)(n /2)
d1 = data[mid]
d2 = data[mid+1]
median = (d1 + d2)/2
median

7.0

### using statistics module

In [46]:
data = [6, 7, 5, 7, 7, 8, 7, 6, 9, 7, 4, 10, 6, 8, 8, 9, 5, 6, 4, 8]
st.median(data)

7.0

### Standard Deviation

#### variance ($\sigma^{2}$) = (1/N) * ${\sum_{i=0}^N ({x_i - \bar x}})^2$ 
 
#### standard deviation ($\sigma$) = $\sqrt{variance}$


In [117]:
def sumAndSquare(x, midVal):
    sum = 0
    for val in x:
        presentVal = (val - midVal)
        presentVal = presentVal **2
        sum = sum + presentVal
        yield sum

In [120]:
result = 0
for x in sumAndSquare(data, mean):
    result = x
result

50.550000000000004

In [122]:
varriance = result / n
varriance

2.5275000000000003

In [123]:
std = math.sqrt(varriance)
std

1.5898113095584647

#### Problem Statement 2:
##### The number of calls from motorists per day for roadside service was recorded for a particular month: 
28, 122, 217, 130, 120, 86, 80, 90, 140, 120, 70, 40, 145, 113, 90, 68, 174, 194, 170, 100, 75, 104, 97, 75, 123, 100, 75, 104, 97, 75, 123, 100, 89, 120, 109

In [264]:
data = [28, 122, 217, 130, 120, 86, 80, 90, 140, 120, 70, 40, 145, 113, 90, 68, 174, 194, 170, 100, 75, 104, 97, 75, 123, 100, 75, 104, 97, 75, 123, 100, 89, 120, 109]

In [252]:
len(data)

35

In [289]:
def Mean(data):
    sum = 0
    for val in data:
        sum = sum + val
    return sum/len(data)        

In [290]:
Mean(data)

107.51428571428572

In [263]:
def mode(data):
    count = 0
    # keep the count of max frequency
    maxCount = 0
    item = 0
    length = len(data)
    # sort the data
    data.sort()
    # iterate
    index = 0
    for index in range(length):
        if( ((index+1) != length) and (data[index] == data[index+1]) ):
            count = count + 1;
            if(count > maxCount):
                maxCount = count
                item = data[index]
        else :            
            count = 0
            count = count + 1
            
    return maxCount, item
        
        

In [256]:
count, val = mode(data)
print(count, val)

4 75


#### Mode = 75

In [257]:
# verifying using staticstics module
st.mode(data)

75

In [275]:
def Median(data):
    # sort the data
    data.sort()
    #get the size
    size = len(data)
    # if odd length
    if(size % 2 != 0):
        # get the mid index
        # get the index
        mid = math.floor(size/2)
        return data[mid]
    else:
        # get mid and mid+1 data
        mid = size/2
        nxt = mid+1
        d1 = data[mid]
        d2 = data[nxt]
        return (d1 + d2)/2.0

In [276]:
print(Median(data))

100


#### Median = 100

In [274]:
# comparing with the statistics module
st.median(data)

100

In [337]:
# calculate varriance
def Varriance(data, Mean, str):
    sum = 0
    for val in data:
        presentVal = (val - Mean)
        presentVal = math.pow(presentVal,2)
        sum = sum + presentVal
    if(str == 'population'):
        return sum/len(data)
    else:
        return sum/(len(data)-1)

In [318]:
print(Varriance(data, Mean(data), 'population'))

1503.3355102040816


In [321]:
# varriance for sample
var = Varriance(data, Mean(data), 'sample')
var

1547.5512605042018

In [322]:
std = math.sqrt(var)
std

39.33892805484412

### standard deviation = 39.33892805484412

In [324]:
# comparing the result with the statistics module
st.variance(data)

1547.5512605042018

In [325]:
st.stdev(data)

39.33892805484412

### Problem Statement 3:
#### The number of times I go to the gym in weekdays, are given below along with its associated probability: 
x = 0, 1, 2, 3, 4, 5 
f(x) = 0.09, 0.15, 0.40, 0.25, 0.10, 0.01 
Calculate the mean no. of workouts in a week. Also evaluate the variance involved in it.

Let x = # of workouts in a week
    E(x) = expected value of x
    x     p(x)
    0     0.09
    1     0.15
    2     0.40
    3     0.25
    4     0.10
    5     0.01

In [328]:
#E(X) = x * p(x)
#E(X) = 0*0.09 + 1 * 0.15 + 2*0.40 + 3*0.25 + 4*0.10 + 5*0.01
#     = 2.15
# mean no. of workouts in a week = 2.15
mean = 2.15

In [342]:
p = [0.09, 0.15, 0.40, 0.25, 0.10, 0.01]
variance = (0-2.15)**2 * 0.09 + (1 - 2.15)**2 *0.15 + (2 - 2.15)**2 *0.40 + (3 - 2.15)**2 *0.25 + (4 - 2.15)**2 *0.10 + (5 - 2.15)**2 *0.01
variance

1.2275

#### Problem Statement 4:
###### Let the continuous random variable D denote the diameter of the hole drilled in an aluminum sheet. 
#### The target diameter to be achieved is 12.5mm. 
#### Random disturbances in the process often result in inaccuracy.
Historical data shows that the distribution of D can be modelled by the 
PDF (𝑑) = 20𝑒−20(𝑑−12.5), 𝑑 ≥ 12.5. 
#### If a part with diameter > 12.6 mm needs to be scrapped,what is the proportion of those parts? 
#### What is the CDF when the diameter is of 11 mm? 
#### What is your conclusion regarding the proportion of scraps?

#### This is an example to Continuous Random variable 
$\int_{-\infty}^{\infty}$ f(x) dx

###### we know that area under curve = 1
$\int_a^b$    where a = 12.5 and b = 12.6

In [370]:
#f(x) = 20𝑒−20(x−12.5)
# on solving we got area b/w 12.5 and 12.6 = 0.865
# Diameter greater than 12.6 need to be scrapped, so 1 - 0.865 = 0.135 

In [371]:
# What is the CDF when the diameter is of 11 mm?
# f(x) will become very very large, that tends to be an infinity

##### Problem Statement 5:
A company manufactures LED bulbs with a faulty rate of 30%. If I randomly select 6
chosen LEDs, what is the probability of having 2 faulty LEDs in my sample?
Calculate the average value of this process. Also evaluate the standard deviation
associated with it.

#### Solution
p = probability of fault = 0.3
q = 1 - p = 0.7
n = no of samples = 6
Either select or not select, so this problem falls under Binomail Distribution's bucket

    p(r) = nCr $p^r$ $q^n-r$    ---- (1)

r = 2

putthing these values in (1), we got
p(2) = 0.11

Mean = np = 1.8
variance = npq = 1.26

std = $\sqrt{variance}$ = 1.1225

#### Problem Statement 6:

Gaurav and Barakha are both preparing for entrance exams. Gaurav attempts to
solve 8 questions per day with a correction rate of 75%, while Barakha averages
around 12 questions per day with a correction rate of 45%. 

What is the probability that each of them will solve 5 questions correctly? 

What happens in cases of 4 and 6 correct solutions? 

What do you infer from it? What are the two main governing factors affecting their ability to solve questions correctly? 

Give a pictorial representation of the same to validate your answer.

#### Problem Statement 7:
Customers arrive at a rate of 72 per hour to my shop. 

What is the probability of 𝑘 customers arriving in 4 minutes? 

a) 5 customers, 

b) not more than 3 customers, 

c) more than 3 customers. Give a pictorial representation of the same to validate your
answer.

In [351]:
# rate = 72 per hour. This is nothing but an interval, so it falls into the category of Poisson's Distribution
# 1.2 customer per min
mean = 1.2 * 4 # in 4 minutes
# poisson's distribution p(x) = e-m * (m)**x / x!
p = math.exp(-4.8) * math.pow(4.8,5) / math.factorial(5)
p

0.17474768364388282

### probability of arriving 5 customers per 4 mins  = 17.47%

In [353]:
#(b) not more than 3 customers
# p(x) = p(0) + p(1) + p(2) + p(3)
p0 = math.exp(-4.8) * math.pow(4.8,0) / math.factorial(0)
p1 = math.exp(-4.8) * math.pow(4.8,1) / math.factorial(1)
p2 = math.exp(-4.8) * math.pow(4.8,2) / math.factorial(2)
p3 = math.exp(-4.8) * math.pow(4.8,3) / math.factorial(3)

In [354]:
p = p0 + p1 + p2 + p3
p

0.2942299164965641

### probability of arriving not more than 3 customers per 4 mins = 29.42%

In [358]:
# more than 3 customers
p4 = 1 - p
p4

0.705770083503436

#### probability of arriving more than 3 customers per 4 mins = 70.78%

#### Problem Statement 8:

I work as a data analyst in Aeon Learning Pvt. Ltd. After analyzing data, I make
reports, where I have the efficiency of entering 77 words per minute with 6 errors per hour. 

What is the probability that I will commit 2 errors in a 455-word financial report?

What happens when the no. of words increases/decreases (in case of 1000 words,255 words)?

How is the 𝜆 affected?

How does it influence the PMF? PMF means Probability Mass Function

Give a pictorial representation of the same to validate your answer.

Given:

77 words/min --- (1)

error rate = 6 errors/hr   ---(2)

From (1)

77*60 = 4620 words/hr

from (2)

6 errors/hr = 6/4620 errors/words = 0.0013

In [360]:
# What is the probability that I will commit 2 errors in a 455-word financial report?
# in an interval of 455 words 
u = 455 * 0.0013
p2 = math.exp(-u) * math.pow(u,2) / math.factorial(2)
p2

0.09682651841129084

In [368]:
# What happens when the no. of words increases/decreases (in case of 1000 words,255 words)?
u = 1000 * 0.0013
p1000 = math.exp(-u) * math.pow(u,2) / math.factorial(2)

u = 255 * 0.0013
p255 = math.exp(-u) * math.pow(u,2) / math.factorial(2)

print(p1000, p255)

0.23028936511374065 0.03944286463025426


In [369]:
# How is the 𝜆 affected?
𝜆255 = 0.3315
𝜆1000 = 1.3
# As no of words increases, mean value of getting error introduced is more

### Problem Statement 9:

Let the continuous random variable D denote the diameter of the hole drilled in an aluminum sheet. The target diameter to be achieved is 12.5mm. 

Random disturbances in the process often result in inaccuracy.

Historical data shows that the distribution of D can be modelled by the PDF, f(d) = 20e−20(d−12.5), d ≥ 12.5.

If a part with diameter > 12.6 mm needs to be scrapped,

what is the proportion of those parts? What is the CDF when the diameter is of 11 mm?

What is the conclusion of this experiment?

### This is already solved as a part of Problem 4

## Problem Statement 10

In [372]:
# Problem Statement 10:
# Please compute the following:
# a) P(Z > 1.26), . P(Z < −0.86), P(Z > −1.37), P(−1.25 < Z < 0.37), . P(Z ≤ −4.6)
# b) Find the value 𝑧 such that 𝑃(𝑍 > 𝑧) = 0.05
# c) Find the value of 𝑧 such that 𝑃(−𝑧 < 𝑍 < 𝑧) = 0.99

# Solution

# (a) From Z-score table, 
#z value of 1.26 = 0.8962
# since, the area under curve = 1
# P(Z > 1.26) = 1 - 0.8962 = 0.1038

# z value of -0.86 = 0.8051
# since, the area under curve = 1
# P(Z < -0.86) = 1 - 0.8051 = 0.1949

# z value of -1.25 = 0.1056
# z value of 0.37  = 0.6443
# since, the area under curve = 1
# P(−1.25 < Z < 0.37) = p(z < 0.37) - p(z>-1.25)
#   = 0.6443 - 0.1056 = 0.5387

# P(Z ≤ −4.6)
# the value of Z ranges for -3.4 to 3.4, 
# so, P(Z ≤ −4.6) ~1

# b) Find the value 𝑧 such that 𝑃(𝑍 > 𝑧) = 0.05
# Solution
# 0.05 is the area of the curve. Its corresponding Z value from the table is 1.64
# The value of z = 1.64

# c) Find the value of 𝑧 such that 𝑃(−𝑧 < 𝑍 < 𝑧) = 0.99
# Solution
# 0.99 is the area of the curve. Its corresponding Z value from the table is 2.33
# so -2.33 < z < 2.33

#### Problem Statement 11:

The current flow in a copper wire follow a normal distribution with a mean of 10 𝑚A and a variance of 4 (𝑚𝐴)2.

What is the probability that a current measurement will exceed 13 𝑚𝐴? 

What is the probability that a current measurement is between 9 and 11mA? 

Determine the current measurement which has a probability of 0.98.

In [377]:
mean = 10
𝜎 = math.sqrt(4)
# It follows normal distribution
# z = (x - mean)/𝜎
z = (13 - 10)/𝜎
# p(x>13) = p(z>1.5)
# from z-table, p(1.5) = 0.9332
# p(z>1.5) = 1 - 0.9332
p = 1 - 0.9332
p

0.06679999999999997

In [378]:
# What is the probability that a current measurement is between 9 and 11mA?
# p(9 < x < 11) => p(-0.5 < z < 0.5)
# p = p(z<0.5) - p(-0.5<z) 
# => p = 0.6915 - 0.3830 = 0.3830

In [379]:
# Determine the current measurement which has a probability of 0.98.
# from z-table z=2.08 has the probability of 0.98
# from standard normal value, we know
# Let x being the current
# z = (x - mean)/𝜎
# => 2.08 = (x - 10)/2
x = (2.08 * 2) + 10
x

14.16

### Problem Statement 12:
The shaft in a piston has its diameter normally distributed with a mean of 0.2508 inch and a standard deviation of 0.0005 inch.

The specifications of the shaft are 0.2500 ∓ 0.0015 inch. 

What proportion of shafts are in sync with the specifications? 

If the process is centered so that the mean is equal to the target value of 0.2500, what proportion of shafts conform to the new specifications? 

What is your conclusion from this experiment?

In [381]:
# Given:
# u = 0.2508 inch
# 𝜎 = 0.0005 inch
# (0.2500-0.0015) < x < (0.2500+0.0015) => 0.2485 < x < 0.2515
# Standard normal value is given by
# z = (x - mean)/𝜎
z1 = (.2485 - .2508)/.0005
z2 = (.2515 - .2508)/.0005

# P ( 0.2485 < X <0.2515 ) = P(−4.6 <Z <1.4)
#  p(z < 1.4) - p(z > -4.6)
p = .9192 - 0
p

0.9192

In [382]:
# What proportion of shafts are in sync with the specifications?
0.9192

0.9192

In [384]:
# If the process is centered so that the mean is equal to the target value of 0.2500, what proportion of shafts conform to the new specifications? 
# u = 0.2500
# x lies between 0.2485 < x < 0.2515
# Standard normal value is given by
# z = (x - mean)/𝜎
z1 = (.2485 - .2500)/.0005
z2 = (.2515 - .2500)/.0005

# P ( 0.2485 < X <0.2515 ) = P(−3 <Z < 3)
# p = p(z < 3) - p(z>-3)
p = 0.9987 - 0.0013
p

0.9974000000000001

##### Conclusion

As the mean is moving towards the targeted value,the probability is moving towards 1